### ISSUE TITLE
BadWindow (invalid Window parameter)

### ISSUE THREAD

**Issue by c-edw** [ggez/ggez#361]
When I run `cargo run --example astroblasto`, I get the following.

```
Adding path "/home/cedwards/Documents/git/ggez/resources"
X Error of failed request:  BadWindow (invalid Window parameter)
  Major opcode of failed request:  4 (X_DestroyWindow)
  Resource id in failed request:  0x0
  Serial number of failed request:  160
  Current serial number in output stream:  163
```

Distribution: Ubuntu 17.10
Kernel: 4.17-rc2

**Comment by icefoxen** [comment]
Thanks for the report, that's new!  Your version of Ubuntu makes me think you might be affected by https://github.com/ggez/ggez/issues/194 but it generally gives a different error message.

Do you get a window at all, even just blinking up for a moment before vanishing?  Can you run any of the examples from https://github.com/Rust-SDL2/rust-sdl2 ?

**Comment by c-edw** [comment]
@icefoxen No window appears at all. I ran the `demo` example in `rust-sdl2` and it seemed to function. I had a red window open.

**Comment by c-edw** [comment]
Some output of `glxinfo`.

```
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) Kabylake GT1.5
OpenGL core profile version string: 4.5 (Core Profile) Mesa 17.2.8
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.0 Mesa 17.2.8
OpenGL shading language version string: 1.30
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.2 Mesa 17.2.8
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```

**Comment by XavilPergis** [comment]
I am having the same problem.

```
> cargo run --example astroblasto
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/examples/astroblasto`
Adding path "/home/xavil/code/rust/ggez/resources"
X Error of failed request:  BadWindow (invalid Window parameter)
  Major opcode of failed request:  4 (X_DestroyWindow)
  Resource id in failed request:  0x0
  Serial number of failed request:  160
  Current serial number in output stream:  161
```

Distro: `Ubuntu 17.10`
Kernel: `4.13.0-39-generic`
Some `glxinfo`:
```
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Intel Open Source Technology Center (0x8086)
    Device: Mesa DRI Intel(R) Haswell Mobile  (0x416)
    Version: 17.2.8
    Accelerated: yes
    Video memory: 1536MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 3.0
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.1

...snip...

OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) Haswell Mobile 
OpenGL core profile version string: 4.5 (Core Profile) Mesa 17.2.8
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
```

**Comment by troiganto** [comment]
Having the same problem running another game using ggez 0.4.2 from crates.io. No window whatsoever appears.

```
X Error of failed request:  BadWindow (invalid Window parameter)
  Major opcode of failed request:  4 (X_DestroyWindow)
  Resource id in failed request:  0x0
  Serial number of failed request:  169
  Current serial number in output stream:  170
```

Distro: `Ubuntu 17.10`
Kernel: `4.13.0-39-generic`

glxinfo is:

```
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Intel Open Source Technology Center (0x8086)
    Device: Mesa DRI Intel(R) Bay Trail  (0xf31)
    Version: 17.2.8
    Accelerated: yes
    Video memory: 1536MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.2
    Max compat profile version: 3.0
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) Bay Trail 
OpenGL core profile version string: 4.2 (Core Profile) Mesa 17.2.8
OpenGL core profile shading language version string: 4.20
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
```

**Comment by buckle2000** [comment]
Same here.
Ubuntu 17.10, kernel 4.13.0-41-generic

Might related to https://github.com/ClioPatria/ClioPatria/issues/46 ?

**Comment by buckle2000** [comment]
```shell
$ git bisect log
git bisect start
# bad: [f3592d613fe95debe32105ed70c939094070376a] Merge pull request #262 from dannyfritz/guide
git bisect bad f3592d613fe95debe32105ed70c939094070376a
# bad: [c0332b357e3268047a37917b3b24c672a762432e] Fixed sRGB colors
git bisect bad c0332b357e3268047a37917b3b24c672a762432e
# good: [1aa07901a955953e66a827344c6fce32bfd7982a] Version bump to 0.2.2
git bisect good 1aa07901a955953e66a827344c6fce32bfd7982a
# good: [fcd9a7d97f469b83c0b177c3964505247be59683] Merge pull request #25 from DenialAdams/master
git bisect good fcd9a7d97f469b83c0b177c3964505247be59683
# skip: [b6e9603a73b404e3ff239dd8d3d25f6da5eb1e2a] Made font drawing work again.
git bisect skip b6e9603a73b404e3ff239dd8d3d25f6da5eb1e2a
# skip: [97b7bff00dfcd1c5366b5fc96675f5953b884575] Merge pull request #45 from vickenty/gfx-init
git bisect skip 97b7bff00dfcd1c5366b5fc96675f5953b884575
# skip: [96c652c133b4f3c7a71e53d9759b18911c4298e9] Made
git bisect skip 96c652c133b4f3c7a71e53d9759b18911c4298e9
# skip: [9028ff65e0828ef3098be1e1127a0a106b7318cd] Implementing color mods.
git bisect skip 9028ff65e0828ef3098be1e1127a0a106b7318cd
# skip: [2ce93812ff9669d53574acbfc9251f584554a6cd] Make font_from_bytes public
git bisect skip 2ce93812ff9669d53574acbfc9251f584554a6cd
# skip: [7ed67ca3ce05c2e4de3b3c0cdb97281446f1a0a8] Draw lines using lyon tessellation.
git bisect skip 7ed67ca3ce05c2e4de3b3c0cdb97281446f1a0a8
# skip: [4570ed6ea320fc84dc3128f4dbaa9c048af6752e] adding std Error implementation for GameError
git bisect skip 4570ed6ea320fc84dc3128f4dbaa9c048af6752e
# bad: [4570ed6ea320fc84dc3128f4dbaa9c048af6752e] adding std Error implementation for GameError
git bisect bad 4570ed6ea320fc84dc3128f4dbaa9c048af6752e
# skip: [3c046c5066449a1201fb12b550870b9de0a0a794] Updated unit test for power-of-2 image scaling
git bisect skip 3c046c5066449a1201fb12b550870b9de0a0a794
# bad: [64a315b6ca1d98fefab3b4163dd538ab4a892bbe] Oh, and a version bump, derp
git bisect bad 64a315b6ca1d98fefab3b4163dd538ab4a892bbe
# skip: [7a4ad633294f9e2469c177fff2134d64b9f87a81] Merge pull request #61 from Gordon-F/master
git bisect skip 7a4ad633294f9e2469c177fff2134d64b9f87a81
# good: [ae22c609f0f35b416a4c93bb01ff1d5031aa8a98] Folded game.rs into event.rs
git bisect good ae22c609f0f35b416a4c93bb01ff1d5031aa8a98
# bad: [8860928dd62182b1aa1038c2c18746632d0108cd] Version bump rodio to 0.5
git bisect bad 8860928dd62182b1aa1038c2c18746632d0108cd
# skip: [896426f76e057a3f94e19c3033b3e5ed3c77720b] Implemented shear.
git bisect skip 896426f76e057a3f94e19c3033b3e5ed3c77720b
# skip: [9b9826bd8015781a4a286662fa0628309298b05c] Merge pull request #52 from vickenty/gfx-cleanup
git bisect skip 9b9826bd8015781a4a286662fa0628309298b05c
# skip: [e4e9ae58d312f3213743e1dda353af1d0c5cfc5f] Messing about with making ZipArchive's more convenient
git bisect skip e4e9ae58d312f3213743e1dda353af1d0c5cfc5f
# bad: [0094267177f0121bf5f879418ffaa91efb713e4a] Implemented get_renderer_info()
git bisect bad 0094267177f0121bf5f879418ffaa91efb713e4a
# skip: [09ab07caa1d09579306f83609286e58f22d94fcd] Trying to figure out why filesystem gives errors on Appveyor
git bisect skip 09ab07caa1d09579306f83609286e58f22d94fcd
# skip: [a520fc37cc7ce900eb542cb70600a703af78c0e2] Totally revamped audio to use rodio more nicely.
git bisect skip a520fc37cc7ce900eb542cb70600a703af78c0e2
# skip: [09df33c0357d4f31c27672247abaab360e790c09] Basic shader and pipeline setup should now work.
git bisect skip 09df33c0357d4f31c27672247abaab360e790c09
# bad: [8484ef6d31a36eb44b563b5b872f68bd6ea74499] Turns out we don't need to depend on gfx_core explicitly.
git bisect bad 8484ef6d31a36eb44b563b5b872f68bd6ea74499
# skip: [937f7b297646e9365b3875bf40e90b710f36b22c] Stupid version bumps for gfx
git bisect skip 937f7b297646e9365b3875bf40e90b710f36b22c
# good: [04f3f1c33245d6883ec5d15d23a47697525577d0] Starting to work on gfx graphics.
git bisect good 04f3f1c33245d6883ec5d15d23a47697525577d0
# bad: [59162221ea6afee74968bf779ea3054cc1aed4f4] Okay it appears to work
git bisect bad 59162221ea6afee74968bf779ea3054cc1aed4f4
# only skipped commits left to test
# possible first bad commit: [59162221ea6afee74968bf779ea3054cc1aed4f4] Okay it appears to work
# possible first bad commit: [937f7b297646e9365b3875bf40e90b710f36b22c] Stupid version bumps for gfx
```

**Comment by buckle2000** [comment]
Error messages:
```
[8484ef6d31a36eb44b563b5b872f68bd6ea74499]

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Program(Vertex(CompilationFailed("0:1(10): error: GLSL 1.50 is not supported. Supported versions are: 1.10, 1.20, 1.30, 1.00 ES, 3.00 ES, 3.10 ES, and 3.20 ES\n")))', libcore/result.rs:945:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

[59162221ea6afee74968bf779ea3054cc1aed4f4]

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Program(Vertex(CompilationFailed("0:1(10): error: GLSL 1.50 is not supported. Supported versions are: 1.10, 1.20, 1.30, 1.00 ES, 3.00 ES, 3.10 ES, and 3.20 ES\n")))', libcore/result.rs:945:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by buckle2000** [comment]
screenshot at [f3592d613fe95debe32105ed70c939094070376a] current master:
![scr1](https://user-images.githubusercontent.com/10850402/40199080-fb35b5d0-5a4a-11e8-8ddd-c9bc7b2f8e8c.png)

screenshot at [fcd9a7d97f469b83c0b177c3964505247be59683]:
![scr2](https://user-images.githubusercontent.com/10850402/40199261-82984c7c-5a4b-11e8-935e-0d9700abae12.png)

**Comment by icefoxen** [comment]
> I noticed some differences...

We switched from using `sdl2` for rendering to using `gfx-rs`, writing our own renderer in the process.  We switched from using `sdl-ttf` for font drawing to using `rusttype`.  We added multisampling.  Commit fcd9a7d97f469b83c0b177c3964505247be59683 was almost a year and a half ago; it's been a busy time!

> Is it a hardware problem? I don't think so.

It's a driver problem with Mesa 17.2 and 17.3, see [est31's comment in this thread](https://github.com/ggez/ggez/issues/194#issuecomment-341443975).
